File: windows-tauri/ui/startup.js
/* LETSCUBE Windows startup screen.
 *
 * Rule for this file: nothing is displayed that was not computed somewhere
 * real. Every fingerprint digit on the screen comes out of `snapshot.peer`,
 * which the shell fills from the certificate its own HTTPS request observed. A
 * snapshot without `peer` — which is every build that has not yet been taught
 * to send it — renders no fingerprint at all rather than a placeholder shaped
 * like one.
 */

const STARTUP_EVENT = "letscube://startup-state";
const stageOrder = [
  "network_check",
  "tls_origin_check",
  "update_check",
  "production_navigation",
  "workspace_ready",
  "complete",
];
const stageLabels = {
  boot: "Готовим подключение",
  network_check: "Проверяем доступ к сети",
  tls_origin_check: "Проверяем сертификат узла",
  update_check: "Смотрим, есть ли обновление",
  production_navigation: "Открываем рабочее пространство",
  workspace_ready: "Рабочее пространство готово",
  complete: "Соединение с узлом подтверждено",
  recoverable_error: "Подключение остановлено",
  // Declared in StartupStage (src-tauri/src/startup.rs) but not currently
  // reachable: nothing calls require_critical_update. It is listed so that an
  // unmapped stage cannot fall back to a message describing a different one.
  critical_update_required: "Требуется обновление",
};

// What the left panel can honestly say before the shell sends a certificate.
// These describe the shell's own request, which is the only thing that has
// been checked at each point.
const shellCheckNotes = {
  boot: "проверка не начиналась",
  network_check: "проверяем доступ к сети",
  tls_origin_check: "проверяем цепочку сертификатов",
  update_check: "цепочка проверена",
  production_navigation: "цепочка проверена",
  workspace_ready: "цепочка проверена",
  complete: "цепочка проверена",
  recoverable_error: "проверка не завершена",
  critical_update_required: "цепочка проверена",
};

const handshake = document.querySelector("#startup-handshake");
const seal = document.querySelector("#startup-center-seal");
const status = document.querySelector("#startup-status");
const failure = document.querySelector("#startup-failure");
const error = document.querySelector("#startup-error");
const retry = document.querySelector("#startup-retry");
const mismatch = document.querySelector("#startup-mismatch");
const mismatchDetail = document.querySelector("#startup-mismatch-detail");
const mismatchRecheck = document.querySelector("#startup-mismatch-recheck");
const mismatchContinue = document.querySelector("#startup-mismatch-continue");
const clientIdentity = document.querySelector('[data-testid="startup-client-fingerprint"]');
const serverIdentity = document.querySelector('[data-testid="startup-server-fingerprint"]');
const dragRegion = document.querySelector("#startup-drag-region");
const minimize = document.querySelector("#startup-window-minimize");
const maximize = document.querySelector("#startup-window-maximize");
const close = document.querySelector("#startup-window-close");

function invokeWindowCommand(command) {
  const invoke = window.__TAURI_INTERNALS__?.invoke;
  return typeof invoke === "function" ? invoke(command).catch(() => undefined) : Promise.resolve();
}

/* A SHA-256 is 32 bytes. Anything that is not exactly 32 bytes of hex is not a
 * fingerprint this screen knows how to show, so it is refused rather than
 * printed in a shape that would look authoritative. */
function normalizeDigest(value) {
  if (typeof value !== "string") return null;
  const hex = value.replace(/[\s:-]/g, "").toUpperCase();
  if (!/^[0-9A-F]{64}$/.test(hex)) return null;
  return hex.match(/../g);
}

/* Two lengths, for two different jobs.
 *
 * While things are going well nobody is comparing anything, and 32 bytes of
 * hex across a splash screen is noise: the leading four bytes are enough to
 * recognise, and the ellipsis says outright that this is a prefix rather than
 * the value. The full 32 bytes appear only where someone actually has to
 * compare two of them — the mismatch state — laid out as four lines of eight,
 * which is the four-line block the identity row has always reserved. */
function digestLines(value, full) {
  const bytes = normalizeDigest(value);
  if (!bytes) return null;
  if (!full) return [`${bytes.slice(0, 4).join(":")}…`];
  return [0, 8, 16, 24].map((offset) => bytes.slice(offset, offset + 8).join(":"));
}

function renderIdentity(node, label, lines, note) {
  const labelNode = node.querySelector("[data-fingerprint-label]");
  const valueNode = node.querySelector("[data-fingerprint-value]");
  labelNode.textContent = label;
  valueNode.replaceChildren();
  valueNode.dataset.digest = lines ? (lines.length === 1 ? "short" : "full") : "none";
  if (lines) {
    // One <span> per line of the digest, and none for anything else, so
    // counting the spans in this block counts fingerprint lines. The e2e
    // geometry check relies on that: four lines or none, never a partial value.
    valueNode.dataset.kind = "digest";
    for (const line of lines) {
      const span = document.createElement("span");
      span.textContent = line;
      valueNode.append(span);
    }
    node.setAttribute("aria-label", `${label}: ${lines.join(":")}`);
    return;
  }
  valueNode.dataset.kind = "note";
  valueNode.textContent = note;
  node.setAttribute("aria-label", `${label}: ${note}`);
}

/* The two panels only ever carry independently obtained values: the left one
 * is what this computer recorded on an earlier connection, the right one is
 * what arrived on the wire just now. Printing one value into both would be the
 * same theatre with extra steps, so the left panel shows nothing at all when
 * there is no record to compare against. */
function renderPeer(snapshot, full) {
  const peer = snapshot.peer && typeof snapshot.peer === "object" ? snapshot.peer : null;
  const observed = digestLines(peer?.observedSha256, full);
  const expected = digestLines(peer?.expectedSha256, full);

  if (!observed) {
    renderIdentity(clientIdentity, "Проверка оболочки", null, shellCheckNotes[snapshot.stage] ?? "состояние неизвестно");
    renderIdentity(serverIdentity, "Отпечаток узла", null, "недоступен в этой сборке");
    return;
  }

  renderIdentity(serverIdentity, "Отпечаток узла", observed, "");
  if (expected) {
    renderIdentity(clientIdentity, "Ожидаемый отпечаток", expected, "");
  } else {
    renderIdentity(clientIdentity, "Ожидаемый отпечаток", null, "первое подключение, сравнивать не с чем");
  }
}

/* The issuer and the validity date come off the certificate, which means they
 * are supplied by whatever answered the connection. `textContent` keeps them
 * from being markup, and these bounds keep a deliberately long string from
 * pushing the notice — and with it the scene — past the bottom of a 640px
 * window. Anything longer is not a name, it is a lever. */
function boundedField(value, limit) {
  if (typeof value !== "string") return null;
  const cleaned = value.replace(/\s+/g, " ").trim();
  if (!cleaned) return null;
  return cleaned.length > limit ? `${cleaned.slice(0, limit - 1)}…` : cleaned;
}

function describeMismatch(peer) {
  const issuer = boundedField(peer?.issuer, 48);
  const notAfter = boundedField(peer?.notAfter, 24);
  const source = [issuer, notAfter ? `действителен до ${notAfter}` : null]
    .filter(Boolean)
    .join(" · ");
  return [
    "Цепочка доверия проверена средствами Windows и признана доверенной — но сертификат не тот,",
    "что был записан при прошлом подключении. Так выглядит и плановая замена сертификата,",
    "и подмена, поэтому сверьте отпечатки выше.",
    source ? `— ${source}.` : "",
  ].join(" ").trim();
}

function focusWhenRevealed(node, control, shouldShow) {
  const wasHidden = node.hidden;
  node.hidden = !shouldShow;
  // Focus goes to the safe action, and only on the transition into view, so a
  // repeated snapshot does not keep pulling focus back.
  if (shouldShow && wasHidden) control.focus();
}

function renderStartup(snapshot) {
  if (!snapshot || typeof snapshot.stage !== "string" || typeof snapshot.connected !== "boolean") return;
  document.body.dataset.stage = snapshot.stage;
  const connected = snapshot.stage === "complete" && snapshot.connected === true;
  const peerChanged = snapshot.stage === "recoverable_error" && snapshot.errorCode === "peer_changed";
  const hardFailure = snapshot.stage === "recoverable_error" && !peerChanged;

  handshake.classList.toggle("is-connected", connected);
  seal.setAttribute("aria-hidden", String(!connected));
  status.textContent = stageLabels[snapshot.stage] ?? stageLabels.boot;
  // The full digest is shown only where someone has to compare two of them.
  renderPeer(snapshot, peerChanged);

  document.body.dataset.verdict = connected
    ? "verified"
    : peerChanged
      ? "changed"
      : hardFailure
        ? "failed"
        : "pending";

  if (hardFailure) {
    // A certificate the trust store rejects lands here, with a retry and no
    // way past. The override below is only ever offered for a pin that no
    // longer matches, never for a chain that failed to validate.
    error.textContent = snapshot.errorCode === "tls_origin"
      ? "Не удалось подтвердить сертификат LETSCUBE. Соединение не установлено."
      : "Сервер LETSCUBE недоступен. Проверьте подключение и повторите попытку.";
  }
  if (peerChanged) {
    mismatchDetail.textContent = describeMismatch(snapshot.peer);
  }

  retry.disabled = false;
  mismatchRecheck.disabled = false;
  mismatchContinue.disabled = false;
  focusWhenRevealed(failure, retry, hardFailure);
  focusWhenRevealed(mismatch, mismatchRecheck, peerChanged);

  const current = stageOrder.indexOf(snapshot.stage);
  // A changed pin is only discoverable after the handshake completed, so the
  // network and certificate stages did finish. Leaving them blank would
  // contradict the notice directly below them, which says the chain validated
  // and the address matched. Every other stop leaves the track where it was.
  const reached = peerChanged ? stageOrder.indexOf("update_check") : current;
  document.querySelectorAll("[data-stage-name]").forEach((item) => {
    const index = stageOrder.indexOf(item.dataset.stageName);
    item.classList.toggle("is-active", index === current);
    item.classList.toggle("is-done", connected || (reached > index && reached >= 0));
  });
}

function invokeGuarded(control, command) {
  control.disabled = true;
  const invoke = window.__TAURI_INTERNALS__?.invoke;
  if (typeof invoke === "function") {
    void invoke(command).catch(() => { control.disabled = false; });
  } else {
    control.disabled = false;
  }
}

retry.addEventListener("click", () => invokeGuarded(retry, "retry_main"));
mismatchRecheck.addEventListener("click", () => invokeGuarded(mismatchRecheck, "retry_main"));
/* Per-occurrence only. The command records the fingerprint that was accepted so
 * the next connection compares against it; it does not set an "always allow"
 * flag, and it does not relax any TLS validation — the chain has already been
 * validated by the time this control can exist. */
mismatchContinue.addEventListener("click", () => invokeGuarded(mismatchContinue, "startup_accept_peer_change"));

dragRegion.addEventListener("mousedown", (event) => {
  if (event.button === 0) void invokeWindowCommand("startup_start_dragging");
});
dragRegion.addEventListener("dblclick", () => void invokeWindowCommand("startup_toggle_maximize"));
minimize.addEventListener("click", () => void invokeWindowCommand("startup_minimize"));
maximize.addEventListener("click", () => void invokeWindowCommand("startup_toggle_maximize"));
close.addEventListener("click", () => void invokeWindowCommand("startup_close_to_tray"));

window.renderStartup = renderStartup;
window.addEventListener(STARTUP_EVENT, (event) => renderStartup(event.detail));
// The identity blocks are empty in the markup, so the boot state is painted
// once here. StartupState::new() starts at Boot, so this is the shell's real
// starting stage and not an assumption.
renderStartup({ stage: "boot", connected: false, errorCode: null });

File: windows-tauri/src-tauri/src/startup.rs
use crate::updater::{is_critical_stable, UpdateChannel};
use serde::Serialize;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StartupStage {
    Boot,
    NetworkCheck,
    TlsOriginCheck,
    UpdateCheck,
    ProductionNavigation,
    WorkspaceReady,
    Complete,
    RecoverableError,
    CriticalUpdateRequired,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StartupErrorCode {
    Network,
    TlsOrigin,
    /// The chain validated and the origin matched, but the leaf certificate is
    /// not the one recorded on an earlier connection. Deliberately distinct
    /// from `TlsOrigin`: that one means the connection was never trusted and
    /// must not be continued past, while this one is what an ordinary renewal
    /// looks like and may be accepted once, explicitly.
    PeerChanged,
}

/// What the shell's own HTTPS request observed about the node's certificate,
/// and what it had recorded before. Every field is produced by the shell; none
/// of it is ever defaulted, so a screen with no certificate to show renders
/// nothing rather than a placeholder.
#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PeerIdentity {
    /// SHA-256 of the DER leaf certificate, lowercase hex.
    pub observed_sha256: String,
    /// The value recorded on an earlier connection, absent on the first one.
    pub expected_sha256: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StartupSnapshot {
    pub stage: StartupStage,
    pub connected: bool,
    pub error_code: Option<StartupErrorCode>,
    /// Omitted entirely when the shell has nothing measured to report, so a
    /// build that never reaches the TLS stage emits exactly the payload it
    /// emitted before this field existed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub peer: Option<PeerIdentity>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct StartupTransitionError {
    pub from: StartupStage,
    pub to: StartupStage,
}

#[derive(Clone, Debug)]
pub struct StartupState {
    stage: StartupStage,
    error_code: Option<StartupErrorCode>,
    peer: Option<PeerIdentity>,
    /// The fingerprint the person chose to continue past, for this attempt
    /// only. It is never written to disk as an "always allow": the next
    /// successful connection records the certificate it actually served, so a
    /// later change asks again.
    accepted_peer: Option<String>,
}

impl StartupState {
    pub fn new() -> Self {
        Self {
            stage: StartupStage::Boot,
            error_code: None,
            peer: None,
            accepted_peer: None,
        }
    }

    /// Records what the shell's request observed. Kept apart from the stage
    /// transitions so that carrying an identity can never move the machine.
    pub fn set_peer(&mut self, peer: PeerIdentity) {
        self.peer = Some(peer);
    }

    pub fn peer(&self) -> Option<&PeerIdentity> {
        self.peer.as_ref()
    }

    pub fn accepted_peer(&self) -> Option<&str> {
        self.accepted_peer.as_deref()
    }

    /// Accepts the observed certificate for the next attempt and rearms the
    /// machine. Only reachable from the state the comparison produced, so it
    /// cannot be used to skip a chain that failed to validate — that failure
    /// carries a different error code and never reaches here.
    pub fn accept_peer_change(&mut self) -> Result<(), StartupTransitionError> {
        if self.stage != StartupStage::RecoverableError
            || self.error_code != Some(StartupErrorCode::PeerChanged)
        {
            return Err(StartupTransitionError {
                from: self.stage,
                to: StartupStage::NetworkCheck,
            });
        }
        let Some(observed) = self.peer.as_ref().map(|peer| peer.observed_sha256.clone()) else {
            return Err(StartupTransitionError {
                from: self.stage,
                to: StartupStage::NetworkCheck,
            });
        };

        self.accepted_peer = Some(observed);
        self.stage = StartupStage::NetworkCheck;
        self.error_code = None;
        Ok(())
    }

    /// The node's certificate is not the recorded one. Only reachable from the
    /// stage that performs the comparison, so a caller cannot reach the
    /// override state from anywhere else in the machine.
    pub fn fail_peer_changed(
        &mut self,
        peer: PeerIdentity,
    ) -> Result<(), StartupTransitionError> {
        if self.stage != StartupStage::TlsOriginCheck {
            return Err(StartupTransitionError {
                from: self.stage,
                to: StartupStage::RecoverableError,
            });
        }

        self.stage = StartupStage::RecoverableError;
        self.error_code = Some(StartupErrorCode::PeerChanged);
        self.peer = Some(peer);
        Ok(())
    }

    pub fn transition(&mut self, next: StartupStage) -> Result<(), StartupTransitionError> {
        if !matches!(
            (self.stage, next),
            (StartupStage::Boot, StartupStage::NetworkCheck)
                | (StartupStage::NetworkCheck, StartupStage::TlsOriginCheck)
                | (StartupStage::TlsOriginCheck, StartupStage::UpdateCheck)
                | (
                    StartupStage::UpdateCheck,
                    StartupStage::ProductionNavigation
                )
                | (
                    StartupStage::ProductionNavigation,
                    StartupStage::WorkspaceReady
                )
                | (StartupStage::WorkspaceReady, StartupStage::Complete)
        ) {
            return Err(StartupTransitionError {
                from: self.stage,
                to: next,
            });
        }

        self.stage = next;
        self.error_code = None;
        Ok(())
    }

    pub fn require_critical_update(
        &mut self,
        channel: UpdateChannel,
        mandatory: bool,
        critical: bool,
    ) -> Result<(), StartupTransitionError> {
        if self.stage != StartupStage::UpdateCheck
            || !is_critical_stable(channel, mandatory, critical)
        {
            return Err(StartupTransitionError {
                from: self.stage,
                to: StartupStage::CriticalUpdateRequired,
            });
        }

        self.stage = StartupStage::CriticalUpdateRequired;
        self.error_code = None;
        Ok(())
    }

    pub fn fail(&mut self, error_code: StartupErrorCode) -> Result<(), StartupTransitionError> {
        if self.stage == StartupStage::CriticalUpdateRequired {
            return Err(StartupTransitionError {
                from: self.stage,
                to: StartupStage::RecoverableError,
            });
        }

        self.stage = StartupStage::RecoverableError;
        self.error_code = Some(error_code);
        Ok(())
    }

    pub fn retry(&mut self) -> Result<(), StartupTransitionError> {
        if self.stage != StartupStage::RecoverableError {
            return Err(StartupTransitionError {
                from: self.stage,
                to: StartupStage::NetworkCheck,
            });
        }

        self.stage = StartupStage::NetworkCheck;
        self.error_code = None;
        Ok(())
    }

    pub fn snapshot(&self) -> StartupSnapshot {
        StartupSnapshot {
            stage: self.stage,
            connected: self.stage == StartupStage::Complete,
            error_code: self.error_code,
            peer: self.peer.clone(),
        }
    }
}

impl Default for StartupState {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::updater::UpdateChannel;

    #[test]
    fn startup_only_reaches_connected_after_every_real_stage() {
        let mut state = StartupState::new();
        assert!(state.transition(StartupStage::TlsOriginCheck).is_err());
        state.transition(StartupStage::NetworkCheck).unwrap();
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state.transition(StartupStage::UpdateCheck).unwrap();
        state
            .transition(StartupStage::ProductionNavigation)
            .unwrap();
        state.transition(StartupStage::WorkspaceReady).unwrap();
        state.transition(StartupStage::Complete).unwrap();
        assert_eq!(state.snapshot().stage, StartupStage::Complete);
        assert!(state.snapshot().connected);
    }

    fn identity(observed: &str, expected: Option<&str>) -> PeerIdentity {
        PeerIdentity {
            observed_sha256: observed.to_owned(),
            expected_sha256: expected.map(str::to_owned),
        }
    }

    #[test]
    fn a_changed_pin_is_only_reachable_from_the_stage_that_compares() {
        // Reaching the override state from anywhere else would mean a failure
        // that never validated a chain could present the "continue" control.
        for stage in [
            StartupStage::Boot,
            StartupStage::NetworkCheck,
            StartupStage::UpdateCheck,
            StartupStage::ProductionNavigation,
        ] {
            let mut state = StartupState::new();
            state.stage = stage;
            assert!(state.fail_peer_changed(identity("a", None)).is_err());
        }

        let mut state = StartupState::new();
        state.transition(StartupStage::NetworkCheck).unwrap();
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state
            .fail_peer_changed(identity("aa", Some("bb")))
            .unwrap();
        let snapshot = state.snapshot();
        assert_eq!(snapshot.stage, StartupStage::RecoverableError);
        assert_eq!(snapshot.error_code, Some(StartupErrorCode::PeerChanged));
        assert!(!snapshot.connected);
    }

    #[test]
    fn accepting_a_change_is_refused_for_every_other_failure() {
        // A chain the trust store rejected, or no connection at all, must not be
        // continuable. Only the comparison's own failure may be accepted.
        for code in [StartupErrorCode::Network, StartupErrorCode::TlsOrigin] {
            let mut state = StartupState::new();
            state.transition(StartupStage::NetworkCheck).unwrap();
            // A fingerprint observed on an earlier attempt in this same process,
            // set before the failure. Without it the call is refused for the
            // wrong reason — there is no value to accept — and the stage guard
            // goes untested: measured, deleting that guard entirely left this
            // test green. The realistic shape is exactly this one, a peer seen
            // once and a later attempt whose chain the trust store rejects.
            state.set_peer(identity("observed", Some("recorded")));
            state.fail(code).unwrap();
            assert!(state.accept_peer_change().is_err());
            assert_eq!(state.snapshot().stage, StartupStage::RecoverableError);
            assert!(state.accepted_peer().is_none());
        }
    }

    #[test]
    fn accepting_a_change_arms_one_attempt_and_records_nothing_else() {
        let mut state = StartupState::new();
        state.transition(StartupStage::NetworkCheck).unwrap();
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state
            .fail_peer_changed(identity("observed", Some("recorded")))
            .unwrap();

        state.accept_peer_change().unwrap();
        assert_eq!(state.accepted_peer(), Some("observed"));
        assert_eq!(state.snapshot().stage, StartupStage::NetworkCheck);
        assert_eq!(state.snapshot().error_code, None);

        // The acceptance is for the value that was actually shown. A later
        // change to some third certificate must stop again rather than inherit
        // this decision.
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state
            .fail_peer_changed(identity("third", Some("recorded")))
            .unwrap();
        assert_ne!(state.accepted_peer(), Some("third"));
    }

    #[test]
    fn a_snapshot_without_a_measured_certificate_carries_no_peer_field() {
        // A build that never reaches the comparison must emit exactly the
        // payload it emitted before the field existed, so the screen renders no
        // fingerprint rather than an empty one.
        let state = StartupState::new();
        let json = serde_json::to_string(&state.snapshot()).unwrap();
        assert!(!json.contains("peer"), "{json}");

        let mut measured = StartupState::new();
        measured.set_peer(identity("observed", None));
        let json = serde_json::to_string(&measured.snapshot()).unwrap();
        assert!(json.contains("\"observedSha256\":\"observed\""), "{json}");
        assert!(json.contains("\"expectedSha256\":null"), "{json}");
    }

    #[test]
    fn recoverable_error_never_reports_connected() {
        let mut state = StartupState::new();
        state.fail(StartupErrorCode::Network).unwrap();
        assert_eq!(state.snapshot().stage, StartupStage::RecoverableError);
        assert!(!state.snapshot().connected);
    }

    #[test]
    fn tls_origin_failure_is_retryable_without_connecting() {
        let mut state = StartupState::new();
        state.transition(StartupStage::NetworkCheck).unwrap();
        state
            .fail(StartupErrorCode::TlsOrigin)
            .expect("TLS/origin failures must enter the retryable state");
        assert_eq!(state.snapshot().stage, StartupStage::RecoverableError);
        assert_eq!(
            state.snapshot().error_code,
            Some(StartupErrorCode::TlsOrigin)
        );
        assert!(!state.snapshot().connected);

        state.retry().unwrap();
        assert_eq!(state.snapshot().stage, StartupStage::NetworkCheck);
    }

    #[test]
    fn public_transition_cannot_enter_critical_update_required() {
        let mut state = StartupState::new();
        state.transition(StartupStage::NetworkCheck).unwrap();
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state.transition(StartupStage::UpdateCheck).unwrap();

        assert!(state
            .transition(StartupStage::CriticalUpdateRequired)
            .is_err());
    }

    #[test]
    fn critical_update_gate_requires_a_stable_typed_decision() {
        let mut test_channel_state = state_at_update_check();
        assert!(test_channel_state
            .require_critical_update(UpdateChannel::Test, true, true)
            .is_err());
        assert_eq!(
            test_channel_state.snapshot().stage,
            StartupStage::UpdateCheck
        );

        let mut stable_channel_state = state_at_update_check();
        stable_channel_state
            .require_critical_update(UpdateChannel::Stable, true, true)
            .unwrap();
        assert_eq!(
            stable_channel_state.snapshot().stage,
            StartupStage::CriticalUpdateRequired
        );
        assert!(!stable_channel_state.snapshot().connected);
    }

    #[test]
    fn critical_update_cannot_be_failed_or_retried() {
        let mut state = state_at_update_check();
        state
            .require_critical_update(UpdateChannel::Stable, true, true)
            .unwrap();

        assert!(state.fail(StartupErrorCode::Network).is_err());
        assert!(state.retry().is_err());
        assert_eq!(state.snapshot().stage, StartupStage::CriticalUpdateRequired);
        assert!(!state.snapshot().connected);
    }

    #[test]
    fn retry_recovers_only_from_recoverable_error_without_connecting() {
        let mut state = StartupState::new();
        assert!(state.retry().is_err());
        assert_eq!(state.snapshot().stage, StartupStage::Boot);
        assert!(!state.snapshot().connected);

        state.fail(StartupErrorCode::Network).unwrap();
        state.retry().unwrap();
        assert_eq!(state.snapshot().stage, StartupStage::NetworkCheck);
        assert_eq!(state.snapshot().error_code, None);
        assert!(!state.snapshot().connected);
    }

    fn state_at_update_check() -> StartupState {
        let mut state = StartupState::new();
        state.transition(StartupStage::NetworkCheck).unwrap();
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state.transition(StartupStage::UpdateCheck).unwrap();
        state
    }

    #[test]
    fn snapshots_use_camel_case_fields_and_snake_case_stages() {
        let mut state = StartupState::new();
        state.transition(StartupStage::NetworkCheck).unwrap();
        state.transition(StartupStage::TlsOriginCheck).unwrap();
        state.transition(StartupStage::UpdateCheck).unwrap();
        state
            .transition(StartupStage::ProductionNavigation)
            .unwrap();
        state.transition(StartupStage::WorkspaceReady).unwrap();
        state.transition(StartupStage::Complete).unwrap();

        assert_eq!(
            serde_json::to_value(state.snapshot()).unwrap(),
            serde_json::json!({
                "stage": "complete",
                "connected": true,
                "errorCode": null,
            })
        );
    }
}
